pkg/server/bundle/client/manager.go

```diff
@@ -39,40 +48,58 @@ func NewManager(config ManagerConfig) *Manager {
 		config.newBundleUpdater = NewBundleUpdater
 	}
 
-	var updaters []BundleUpdater
+	updaters := make(map[string]BundleUpdater)
 	for trustDomain, trustDomainConfig := range config.TrustDomains {
-		updaters = append(updaters, config.newBundleUpdater(BundleUpdaterConfig{
+		updaters[trustDomain] = config.newBundleUpdater(BundleUpdaterConfig{
 			TrustDomainConfig: trustDomainConfig,
 			TrustDomain:       trustDomain,
 			Log:               config.Log.WithField("trust_domain", trustDomain),
 			DataStore:         config.DataStore,
-		}))
+		})
 	}
 
 	return &Manager{
+		log:      config.Log,
 		clock:    config.Clock,
 		updaters: updaters,
 	}
 }
 
 func (m *Manager) Run(ctx context.Context) error {
 	var tasks []func(context.Context) error
-	for _, updater := range m.updaters {
+	for trustDomain, updater := range m.updaters {
 		tasks = append(tasks, func(ctx context.Context) error {
-			return m.runUpdater(ctx, updater)
+			return m.runUpdater(ctx, trustDomain, updater)
 		})
 	}
 
 	return util.RunTasks(ctx, tasks...)
 }
 
-func (m *Manager) runUpdater(ctx context.Context, updater BundleUpdater) error {
-	var refreshHint time.Duration
+func (m *Manager) runUpdater(ctx context.Context, trustDomain string, updater BundleUpdater) error {
 	for {
-		if r, err := updater.UpdateBundle(ctx); err == nil {
-			refreshHint = r
+		var nextRefresh time.Duration
+		localBundle, remoteBundle, _ := updater.UpdateBundle(ctx)
+		switch {
+		case remoteBundle != nil:
```

Comment:
Could it be possible that a bundle refresh gets scheduled with a higher refresh hint time than the stored bundle?

I was thinking in the case that `UpdateBundle()` fails to store the endpoint bundle and the endpoint bundle has a refresh hint time considerably higher than the local bundle. 

Perhaps we could get the UpdateBundle() error and:

`case remoteBundle != nil && err == nil:`

Reply:
It's hard to tell what the right thing to do here is since even if you go off of the old bundle, it might not be right either. However, I think it's confusing to return the endpoint bundle if there was no change or if it failed to persist so I'm going to change UpdateBundle to only return the endpoint bundle when it was retrieved AND successfully stored.